src/napari_matplotlib/histogram.py

```diff
@@ -55,6 +109,31 @@ def on_update_layers(self) -> None:
         for layer in self.viewer.layers:
             layer.events.contrast_limits.connect(self._update_contrast_lims)
 
+        if not self.layers:
+            return
+
+        # Reset to bin start, stop and step
+        layer_data = self._get_layer_data(self.layers[0])
+        self.autoset_widget_bins(data=layer_data)
+
+        # Only allow integer bins for integer data
+        # And only allow values greater than 0 for unsigned integers
+        n_decimals = 0 if np.issubdtype(layer_data.dtype, np.integer) else 2
+        is_unsigned = layer_data.dtype.kind == "u"
+        minimum_value = 0 if is_unsigned else -1e10
+
+        # Disable callbacks whilst widget values might change
+        for widget in self._bin_widgets.values():
+            widget.blockSignals(True)
+
+        self._bin_widgets["start"].setDecimals(n_decimals)
+        self._bin_widgets["stop"].setDecimals(n_decimals)
+        self._bin_widgets["start"].setMinimum(minimum_value)
+        self._bin_widgets["stop"].setMinimum(minimum_value)
+
+        for widget in self._bin_widgets.values():
+            widget.blockSignals(False)
```

Comment:
temporarily disable the callbacks otherwise the plot is re-drawn each time the widget value changes